galerautils/src/gu_enc_mmap.cpp

```diff
@@ -0,0 +1,628 @@
+        gu_throw_error(errno) << "install_signal_handler() signal handler installation failed";
+    }
+}
+
+
+EncMMap::EncMMap(const std::string& key, std::shared_ptr<MMap> rawmmap,
+                 size_t cachePageSize, size_t cacheSize, bool syncOnDestroy,
+                 size_t encryptionStartOffset)
+: mmapraw_(rawmmap)
+, pageSize_(cachePageSize)
+, mmaprawPtr_(static_cast<unsigned char*>(mmapraw_->get_ptr()))
+, vMemSize_(mmapraw_->get_size())
+// mmap 2 pages more: 1st for aligning start, 2nd if the last underlying page is not aligned
+, mmap_ptr_(static_cast<unsigned char*>(mmap(nullptr, vMemSize_ + 2*pageSize_, PROT_NONE, MAP_ANONYMOUS|MAP_PRIVATE, -1, 0)))
+, base_(nullptr)
+, memoryManagerP_(memoryManagerPool.allocate(cachePageSize, cacheSize))
+, memoryManager_(*memoryManagerP_)
+, vpage2protectionGuard_()
+, vpage2protection_(nullptr)
+, vpage2ppage_()
+, pagesCnt_(0)
+, mapped_(mmap_ptr_ != MAP_FAILED)
+, lastPageSize_(pageSize_)
+, encryptionStartOffset_(encryptionStartOffset)
+, defaultPageProtection_(PROT_READ | PROT_WRITE )
+, readAheadCnt_(0)
+, lock_(ATOMIC_FLAG_INIT)
+, encryptor_()
+, decryptor_()
+, syncOnDestroy_(syncOnDestroy) {
+
+    if (!mapped_)
+    {
+        gu_throw_error(errno) << "EncMMap::EncMMap() mmap() on anonymous failed";
+    }
+
+    // we need base_ to be aligned with pageSize_ for easier calculations later
+    // here we will loose at most pageSize_ at the beginning
+    base_ = reinterpret_cast<unsigned char*>(((ptr2ull(mmap_ptr_) + pageSize_) / pageSize_) * pageSize_);
+
+    S_DEBUG_A("EncMMap::EncMMap() this: x%llX, mmap_ptr: x%llX aligned mapping: (x%llX - x%llX) (%ld bytes)\n",
+        ptr2ull(this), ptr2ull(mmap_ptr_), ptr2ull(base_), ptr2ull(base_) + vMemSize_, vMemSize_);
+    // install signal handler common for all objects
+    std::call_once(signal_handler_once, install_signal_handler);
+    pagesCnt_ = vMemSize_ / pageSize_;
+    if (vMemSize_ % pageSize_) {
+        // if the size is not aligned, the last page is smaller than pageSize_
+        lastPageSize_ = vMemSize_ % pageSize_;
+        S_DEBUG_A("EncMMap::EncMMap() adding page, size not aligned: %ld, lastPageSize: %ld\n",
+          vMemSize_, lastPageSize_);
+        pagesCnt_++;
+    }
+
+    S_DEBUG_A("EncMMap::EncMMap() allocated pages cnt: %ld\n", pagesCnt_);
+    vpage2protectionGuard_ = std::shared_ptr<int>(new int[pagesCnt_], [](int *p) { delete[] p; });
+    vpage2protection_ = vpage2protectionGuard_.get();
+    addEncMMap(this, base_, vMemSize_);
+
+    // we set up vpage2protection_ map inside
+    set_key(key);
+}
+
+EncMMap::~EncMMap() {
+    S_DEBUG_A("EncMMap::~EncMMap() this: x%llX, mmap_ptr: x%llX aligned mapping: (x%llX - x%llX) (%ld bytes)\n",
+        ptr2ull(this), ptr2ull(mmap_ptr_), ptr2ull(base_), ptr2ull(base_) + vMemSize_, vMemSize_);
+    if (mapped_)
+    {
+        try { unmap(); } catch (Exception& e) { log_error << e.what(); }
+    }
+
+    delEncMMap(this);
```

Comment:
`unmap()` call still needs `this` object to be present in encMMaps container because it will potentially do syncing, which will potentially trigger signals. 
On the other hand, if we receive signal after `munmap` it means that the application is doing something really wrong, because it is destroying MMap object and at the same time accessing it...
I think we can improve it in the way that removing from encMMaps container will happen after `sync()`, but before actual `munmap`.

Reply:
Removal from encMMaps map moved after sync and before actual unmap. It is still up to the application to not use MMapEnc object while deletion/unmap, because it will cause an unhandled segfault anyway.